dbt/adapters/bigquery/dataproc/batch.py

```diff
@@ -0,0 +1,63 @@
+from typing import Union, Dict
+
+import time
+from datetime import datetime
+from google.cloud.dataproc_v1 import (
+    CreateBatchRequest,
+    BatchControllerClient,
+    Batch,
+    GetBatchRequest,
+)
+from google.protobuf.json_format import ParseDict
+
+from dbt.adapters.bigquery.connections import DataprocBatchConfig
+
+_BATCH_RUNNING_STATES = [Batch.State.PENDING, Batch.State.RUNNING]
+DEFAULT_JAR_FILE_URI = "gs://spark-lib/bigquery/spark-bigquery-with-dependencies_2.12-0.21.1.jar"
+
+
+def create_batch_request(batch: Batch, batch_id, project, region) -> CreateBatchRequest:
+    return CreateBatchRequest(
+        parent=f"projects/{project}/locations/{region}",  # type: ignore
+        batch_id=batch_id,
+        batch=batch,  # type: ignore
+    )
+
+
+def poll_batch_job(parent, batch_id, job_client: BatchControllerClient, timeout: int) -> Batch:
+    batch_name = "".join([parent, "/batches/", batch_id])
+    state = Batch.State.PENDING
+    response = None
+    run_time = 0
+    while state in _BATCH_RUNNING_STATES and run_time < timeout:
+        time.sleep(1)
+        response = job_client.get_batch(  # type: ignore
+            request=GetBatchRequest(name=batch_name),  # type: ignore
+        )
+        run_time = datetime.now().timestamp() - response.create_time.timestamp()  # type: ignore
+        state = response.state
+    if not response:
+        raise ValueError("No response from Dataproc")
+    if run_time >= timeout:
+        raise ValueError(
+            f"Operation did not complete within the designated timeout of {timeout} seconds."
+        )
+    if state != Batch.State.SUCCEEDED:
```

Comment:
Should we check happy path first and return if we're there? Is there a scenario (other than finishing successfully .5s after `timeout`) where `state == Batch.State.SUCCEEDED` but we want to raise an error instead of returning `response`?

Reply:
moved the timeout check to be nested under the succeeded check